autoparallel/activation_checkpointing.py

```diff
@@ -66,6 +66,91 @@ def is_wait_tensor_from_fsdp(node: torch.fx.Node) -> bool:
+                    # %getitem_3 : [num_users=1] = call_function[target=operator.getitem](args = (%split, 1), kwargs = {})
+                    # %getitem_4 : [num_users=1] = call_function[target=operator.getitem](args = (%split, 2), kwargs = {})
+                    # %getitem_5 : [num_users=1] = call_function[target=operator.getitem](args = (%split, 3), kwargs = {})
+                    # %cat_1 : [num_users=1] = call_function[target=torch.ops.aten.cat.default](args = ([%getitem_2, %getitem_3, %getitem_4, %getitem_5], 1), kwargs = {})  # noqa: E501
+                    if (
+                        w.op == "call_function"
+                        and w.target == torch.ops.aten.split.Tensor
+                    ):
+                        if all(
+                            split_user.op == "call_function"
+                            and split_user.target == operator.getitem
+                            and len(split_user.users) == 1
+                            for split_user in w.users
+                        ):
+                            getitem_users = list(
+                                next(iter(getitem_node.users))
+                                for getitem_node in w.users
+                            )
+                            potential_cat_op = getitem_users[0]
+                            if all(
+                                potential_cat_op == getitem_user
+                                for getitem_user in getitem_users
+                            ) and (
+                                potential_cat_op.target == torch.ops.aten.cat.default
+                            ):
+                                w = potential_cat_op
+                                continue
+                    break
+                user = next(iter(w.users))
+                if len(user.all_input_nodes) > 1:
+                    break
+                w = user
+            # 4. Stores the last node in this chain as `last_wait_chain_user`
+            last_wait_chain_user = w
+            # 5. Check if the last node in this chain is used in backward
+            is_used_in_backward = False
+            for downstream_user in last_wait_chain_user.users:
+                if _has_tag_is_backward(downstream_user):
+                    is_used_in_backward = True
+                    break
+            if is_used_in_backward:
+                # 6. If the last node in this chain is used in backward, only then we save the wait_tensor
+                nodes_to_save.append(last_ag_wait_node)
+                primal_origins.append(primal)
+    print(f"force_save_fsdp_all_gather, primal_origins: {primal_origins}")
+    print(f"force_save_fsdp_all_gather, nodes_to_save: {nodes_to_save}")
+    for node in nodes_to_save:
+        node.meta["recompute"] = CheckpointPolicy.MUST_SAVE
+        node.meta["ac_graph_id"] = AP_AC_GRAPH_ID
```

Comment:
i'm not familiar enough with AC to know how this is used (saw the comment above about ID being 'large' but i'm not sure why that's good

Reply:
This is copied from @fmassa's code